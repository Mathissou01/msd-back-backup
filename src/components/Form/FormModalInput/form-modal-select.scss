@use "src/styles/01-settings/settings.colors" as colors;

.c-FormModalInput {
  display: flex;
  flex-direction: column;
  gap: 10px;
  align-items: flex-start;

  &__Label {
    color: colors.$blue-2;
    font-weight: 700;
    font-size: 16px;
  }

  &__Input {
    color: colors.$blue-1;
    font-weight: 500;
    font-size: 14px;
    background-color: transparent;
    border: 0;
  }

  &__ModalTitle {
    color: colors.$expert-blue;
    font-weight: 700;
    font-size: 28px;
    line-height: 35px;
  }

  &__ModalSubtitle {
    padding-top: 4px;
    font-weight: 500;
    font-size: 20px;
    line-height: 25px;
  }

  &__ModalRequired {
    padding-top: 16px;
    color: colors.$gray-600;
    font-weight: 400;
    font-size: 14px;
    line-height: 21px;
  }

  &__ModalForm {
    display: flex;
    flex-direction: column;
    gap: 32px;
    align-items: flex-start;
  }

  &__ModalButtons {
    display: flex;
    flex-direction: row;
    gap: 24px;
  }
}
