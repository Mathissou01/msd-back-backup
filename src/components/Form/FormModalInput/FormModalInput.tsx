import {
  FieldValues,
  FormProvider,
  Mode,
  useForm,
  useFormContext,
} from "react-hook-form";
import { Validate } from "react-hook-form/dist/types/validator";
import { FieldErrors } from "react-hook-form/dist/types/errors";
import { ErrorMessage } from "@hookform/error-message";
import React, { ReactNode, useRef } from "react";
import { isTruthyObjectOrArray } from "../../../lib/utilities";
import CommonButton from "../../Common/CommonButton/CommonButton";
import CommonModalWrapper, {
  CommonModalWrapperRef,
} from "../../Common/CommonModalWrapper/CommonModalWrapper";
import CommonErrorText from "../../Common/CommonErrorText/CommonErrorText";
import FormLabel from "../FormLabel/FormLabel";
import "./form-modal-select.scss";

interface IFormModalInputProps<T> {
  name: string;
  label: string;
  secondaryLabel?: string;
  displayTransform: (...args: Array<Partial<T>>) => ReactNode;
  buttonLabel: string;
  modalTitle: string;
  modalSubtitle?: string;
  isRequired?: boolean;
  hasRequiredChildren?: "some" | "all" | null;
  isDisabled?: boolean;
  onValidate?: Validate<T>;
  onSubmit: (data: { [key: string]: Partial<T> }) => void;
  onInvalid?: (errors: Partial<FieldErrors>) => void;
  children: ReactNode;
  formValidationMode: Mode;
}

export default function FormModalInput<T extends FieldValues>({
  name,
  label,
  secondaryLabel,
  displayTransform,
  buttonLabel,
  modalTitle,
  modalSubtitle,
  isRequired = false,
  hasRequiredChildren = null,
  isDisabled = false,
  onValidate,
  onSubmit,
  onInvalid,
  children,
  formValidationMode,
}: IFormModalInputProps<T>) {
  const childRef = useRef<CommonModalWrapperRef>(null);
  const buttonRef = useRef<HTMLButtonElement>(null);

  /* Static Data */
  const childrenRequiredAllLabel = "Tous les champs sont obligatoires";
  const childrenRequiredSomeLabel = "* Champs obligatoires";
  const submitButtonLabel = "Enregistrer";
  const cancelButtonLabel = "Annuler";
  const errorMessages = {
    required: "Ce champ est obligatoire",
  };

  /* Local Data */
  const {
    register,
    watch,
    formState: { isSubmitting, errors },
  } = useFormContext();
  const form = useForm<T>({ mode: formValidationMode });
  const { handleSubmit, reset } = form;
  const currentParentValues = watch(name);

  function handleClose() {
    reset();
    buttonRef.current?.focus();
  }

  function onLocalSubmitValid(data: { [name: string]: Partial<T> }) {
    childRef.current?.toggleModal(false);
    onSubmit(data);
  }

  function onLocalSubmitInvalid(errors: Partial<FieldErrors>) {
    if (onInvalid) onInvalid(errors);
  }

  return (
    <>
      <div className="c-FormModalInput" data-testid="form-modal-input">
        <FormLabel
          forId={`${name}_button`}
          label={label}
          isRequired={isRequired}
          secondaryLabel={secondaryLabel}
        />
        <input
          className="c-FormModalInput__Input"
          {...register(name, {
            required: { value: isRequired, message: errorMessages.required },
            validate: onValidate,
          })}
          type="hidden"
          disabled={true}
        />
        {isTruthyObjectOrArray(currentParentValues)
          ? displayTransform(currentParentValues)
          : ""}
        <div aria-invalid={!!errors[name]} aria-errormessage={`${name}_error`}>
          <CommonButton
            label={buttonLabel}
            onClick={() => childRef.current?.toggleModal(true)}
            isDisabled={isSubmitting || isDisabled}
            formLabelId={`${name}_button`}
            buttonRef={buttonRef}
          />
        </div>
        <ErrorMessage
          errors={errors}
          name={name}
          render={({ message }) => (
            <CommonErrorText message={message} errorId={`${name}_error`} />
          )}
        />
      </div>
      <CommonModalWrapper ref={childRef} onClose={handleClose}>
        <FormProvider {...form}>
          <form
            onSubmit={(event) => {
              event.stopPropagation();
              handleSubmit(onLocalSubmitValid, onLocalSubmitInvalid)(event);
            }}
            className="c-FormModalInput__ModalForm"
          >
            <hgroup>
              <div className="c-FormModalInput__ModalTitle">{modalTitle}</div>
              {modalSubtitle && (
                <div className="c-FormModalInput__ModalSubtitle">
                  {modalSubtitle}
                </div>
              )}
              {hasRequiredChildren !== null && (
                <div className="c-FormModalInput__ModalRequired">
                  {hasRequiredChildren === "some"
                    ? childrenRequiredSomeLabel
                    : hasRequiredChildren === "all"
                    ? childrenRequiredAllLabel
                    : ""}
                </div>
              )}
            </hgroup>
            {children}
            <div className="c-FormModalInput__ModalButtons">
              <CommonButton
                type="submit"
                label={submitButtonLabel}
                style="primary"
              />
              <CommonButton
                type="button"
                label={cancelButtonLabel}
                onClick={() => childRef.current?.toggleModal(false)}
              />
            </div>
          </form>
        </FormProvider>
      </CommonModalWrapper>
    </>
  );
}
