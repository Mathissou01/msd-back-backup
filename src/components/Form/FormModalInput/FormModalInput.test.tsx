import { render, screen } from "@testing-library/react";
import { FormProvider, useForm } from "react-hook-form";
import React, { ReactNode } from "react";
import FormModalInput from "./FormModalInput";

const mock = {
  name: "form-modal-input",
  label: "form modal input",
  buttonLabel: "button label",
  modalTitle: "modal title",
};

const Wrapper = (props: { children: ReactNode }) => {
  const formMethods = useForm();
  return (
    <>
      <FormProvider {...formMethods}>{props.children}</FormProvider>
    </>
  );
};

describe("FormModalInput", () => {
  it("renders", () => {
    const handleTransform = jest.fn();
    const handleSubmit = jest.fn();
    const { container } = render(
      <Wrapper>
        <FormModalInput
          name={mock.name}
          label={mock.label}
          displayTransform={handleTransform}
          buttonLabel={mock.buttonLabel}
          modalTitle={mock.modalTitle}
          onSubmit={handleSubmit}
          formValidationMode={"onChange"}
          isDisabled={false}
        >
          <div />
        </FormModalInput>
      </Wrapper>,
    );

    const modalInput = screen.getByTestId("form-modal-input");
    expect(modalInput).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
});
