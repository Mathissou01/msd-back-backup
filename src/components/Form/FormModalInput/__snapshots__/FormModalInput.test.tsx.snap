// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormModalInput renders 1`] = `
<div>
  <div
    class="c-FormModalInput"
    data-testid="form-modal-input"
  >
    <label
      class="c-FormLabel"
      for="form-modal-input_button"
    >
      <div
        class="c-FormLabel__Label"
      >
        <span>
          form modal input
        </span>
      </div>
    </label>
    <input
      class="c-FormModalInput__Input"
      disabled=""
      name="form-modal-input"
      type="hidden"
    />
    <div
      aria-errormessage="form-modal-input_error"
      aria-invalid="false"
    >
      <button
        class="c-CommonButton c-CommonButton_secondary c-CommonButton_fontSmall"
        id="form-modal-input_button"
        type="button"
      >
        button label
      </button>
    </div>
  </div>
</div>
`;
